src/python/foglamp/core/server_daemon.py

```diff
@@ -75,38 +83,54 @@ def start():
             _start_server()
 
 
-def stop():
-    """Stops the FogLAMP process if it is running"""
+class StopProcessFailed(Exception):
+    pass
 
-    pid = get_pid()
 
-    if pid is None:
+def stop(pid = None):
+    """Stops FogLAMP if it is running
+
+    :param pid: Optional pid of the daemon process
+    """
+
+    if not pid:
+        pid = get_pid()
+
+    if not pid:
         print("FogLAMP is not running")
-        return 
+        return
 
-    # Kill the daemon process
-    # TODO This should time out and throw an exception
+    stopped = False
+    
     try:
-        while True:
+        for retry_index in range(_MAX_STOP_RETRY):
```

Comment:
Python community doesn't agree on using _ for iterating. https://stackoverflow.com/questions/3685974/python-iterate-a-certain-number-of-times-without-storing-the-iteration-number-a